File: scripts/generate-content.ts
/* eslint-disable unicorn/prefer-top-level-await */
import fs from "node:fs"
import path from "node:path"

import type { BaseChatModel } from "@langchain/core/language_models/chat_models"
import {
  JsonOutputParser,
  StringOutputParser,
} from "@langchain/core/output_parsers"
import { PromptTemplate } from "@langchain/core/prompts"
import { RunnableSequence } from "@langchain/core/runnables"
import { ChatMistralAI } from "@langchain/mistralai"
import axios from "axios"
import dotenv from "dotenv"
import yargs from "yargs"

import { prompt as contentCreatorPrompt } from "./prompts/content-creator"
import { prompt as metaGeneratorPrompt } from "./prompts/meta-generator"
import { prompt as outlineBuilderPrompt } from "./prompts/outline-builder"

dotenv.config()

const argv = yargs(process.argv.slice(2))
  .option("n-content", {
    alias: "n",
    describe: "Number of content pieces to generate",
    type: "number",
    default: 1,
  })
  .option("topic", {
    alias: "t",
    describe: "Topic for content generation",
    type: "string",
    default: "stock-trading", // TODO: Change the default topic
  })
  .parseSync()

const BACKGROUND_COLORS = [
  "3498db",
  "2ecc71",
  "e74c3c",
  "f39c12",
  "9b59b6",
  "1abc9c",
  "d35400",
  "c0392b",
  "16a085",
  "2980b9",
  "8e44ad",
  "2c3e50",
]

const THUMBNAIL_WIDTH = 1200
const THUMBNAIL_HEIGHT = 630
const TEXT_COLOR = "ffffff"

type Meta = {
  meta: {
    title: string
    description: string
  }
}

async function main() {
  const count = argv["n-content"]
  const topic = argv["topic"]

  const model = new ChatMistralAI({
    model: "open-mixtral-8x7b",
    temperature: 0.8,
  })

  for (let i = 0; i < count; i++) {
    try {
      console.log(`Generating content ${i + 1} of ${count}...`)

      const outline = await generateOutline(model, topic)
      const content = await generateContent(model, outline)
      const meta = await generateMeta(model, outline)

      await createMarkdownFile(content, meta)
    } catch (error) {
      console.error("An error occurred:", error)
      console.error("Skipping to the next content generation...")
    }

    console.log("Content generation complete. Moving to the next content...")
  }
}

// Helper functions with improved error handling and typing
async function generateOutline(
  model: BaseChatModel,
  topic: string,
): Promise<string> {
  console.log("Generating outline...")
  const outlineBuildChain = RunnableSequence.from([
    PromptTemplate.fromTemplate(outlineBuilderPrompt),
    model,
    new StringOutputParser(),
  ])
  const outline = await outlineBuildChain.invoke({ Topic: topic })
  console.log("Outline generated.")
  return outline
}

async function generateContent(
  model: BaseChatModel,
  outline: string,
): Promise<string> {
  console.log("Generating content based on the outline...")
  const contentCreateChain = RunnableSequence.from([
    PromptTemplate.fromTemplate(contentCreatorPrompt),
    model,
    new StringOutputParser(),
  ])
  const content = await contentCreateChain.invoke({ Outline: outline })
  console.log("Content generated.")
  return content
}

async function generateMeta(
  model: BaseChatModel,
  outline: string,
): Promise<Meta> {
  console.log("Generating meta data based on the outline...")
  const metaGenerateChain = RunnableSequence.from([
    PromptTemplate.fromTemplate(metaGeneratorPrompt),
    model,
    new JsonOutputParser(),
  ])

  const meta = await metaGenerateChain.invoke({ Outline: outline })
  console.log("Meta data generated.")
  return meta as Meta
}

async function generatePlaceholderThumbnail(fileName: string): Promise<string> {
  const text = fileName.replaceAll("-", " ")
  const backgroundColor =
    BACKGROUND_COLORS[Math.floor(Math.random() * BACKGROUND_COLORS.length)]

  const url = `https://dummyimage.com/${THUMBNAIL_WIDTH}x${THUMBNAIL_HEIGHT}/${backgroundColor}/${TEXT_COLOR}.png&text=${encodeURIComponent(text)}`

  const response = await axios({
    url,
    method: "GET",
    responseType: "stream",
  })

  const thumbnailPath = path.join(
    "src",
    "posts",
    "blog",
    "images",
    `${fileName}.png`,
  )
  const writer = fs.createWriteStream(thumbnailPath)

  response.data.pipe(writer)

  return new Promise((resolve, reject) => {
    writer.on("finish", () => resolve(thumbnailPath))
    writer.on("error", reject)
  })
}

async function createMarkdownFile(content: string, meta: Meta): Promise<void> {
  const currentDate = new Date()
    .toLocaleString("en-US", {
      timeZone: "Asia/Seoul",
      year: "numeric",
      month: "2-digit",
      day: "2-digit",
      hour: "2-digit",
      minute: "2-digit",
      hour12: false,
    })
    .replace(/(\d+)\/(\d+)\/(\d+),/, "$3-$1-$2")

  const fileName = meta.meta.title
    .replaceAll(/\s+/g, "-")
    .replaceAll(":", "-")
    .toLowerCase()

  const thumbnailPath = await generatePlaceholderThumbnail(fileName)
  const relativeThumbnailPath = path
    .relative(path.join("src", "posts", "blog"), thumbnailPath)
    .replaceAll("\\", "/")

  const filePath = path.join("src", "posts", "blog", `${fileName}.md`)

  const markdownContent = `---
title: ${meta.meta.title.replaceAll(":", "-")}
category: "constant"
date: ${currentDate} +09:00
desc: ${meta.meta.description.replaceAll(":", "-")}
thumbnail: "./${relativeThumbnailPath}"
alt: "Thumbnail for ${meta.meta.title}"
---

${content}`

  fs.mkdirSync(path.dirname(filePath), { recursive: true })
  fs.writeFileSync(filePath, markdownContent)

  console.log(`Markdown file created: ${filePath}`)
}

main().catch(error => console.error("An error occurred:", error))

File: scripts/generate_content.py
# Logging setup
import logging
import random
from datetime import datetime
from enum import StrEnum
from pathlib import Path
from typing import Annotated, TypedDict

import requests
import typer
from dotenv import load_dotenv
from langchain.prompts import PromptTemplate
from langchain_core.language_models import BaseChatModel
from langchain_core.output_parsers import JsonOutputParser, StrOutputParser
from langchain_mistralai import ChatMistralAI
from langchain_openai import ChatOpenAI

# Constants moved to a separate config.py file
from scripts.config import (
    BACKGROUND_COLORS,
    TEXT_COLOR,
    THUMBNAIL_HEIGHT,
    THUMBNAIL_WIDTH,
)
from scripts.prompts.content_creator import prompt as content_creator_prompt
from scripts.prompts.meta_generator import prompt as meta_generator_prompt
from scripts.prompts.outline_builder import prompt as outline_builder_prompt

load_dotenv()

logging.basicConfig(level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s")
logger = logging.getLogger(__name__)


class MetaItem(TypedDict):
    title: str
    description: str


# Type definitions
class Meta(TypedDict):
    meta: MetaItem


app = typer.Typer()


def generate_outline(model: BaseChatModel, topic: str) -> str:
    logger.info("Generating outline...")

    prompt = PromptTemplate.from_template(outline_builder_prompt)
    output_parser = StrOutputParser()

    chain = prompt | model | output_parser

    try:
        outline = chain.invoke(input={"Topic": topic})
        logger.info("Outline generated successfully.")
        return outline
    except Exception as e:
        logger.error(f"Error generating outline: {e}")
        raise


def generate_content(model: BaseChatModel, outline: str) -> str:
    logger.info("Generating content based on the outline...")

    prompt = PromptTemplate.from_template(content_creator_prompt)
    output_parser = StrOutputParser()
    chain = prompt | model | output_parser

    try:
        content = chain.invoke(input={"Outline": outline})
        logger.info("Content generated successfully.")
        return content
    except Exception as e:
        logger.error(f"Error generating content: {e}")
        raise


def generate_meta(model: BaseChatModel, outline: str) -> Meta:
    logger.info("Generating meta data based on the outline...")

    prompt = PromptTemplate.from_template(meta_generator_prompt)
    output_parser = JsonOutputParser()
    chain = prompt | model | output_parser

    try:
        meta = chain.invoke(input={"Outline": outline})
        logger.info("Meta data generated successfully.")
        return meta
    except Exception as e:
        logger.error(f"Error generating meta data: {e}")
        raise


def generate_placeholder_thumbnail(file_name: str) -> Path:
    text = file_name.replace("-", " ")
    background_color = random.choice(BACKGROUND_COLORS)
    url = f"https://dummyimage.com/{THUMBNAIL_WIDTH}x{THUMBNAIL_HEIGHT}/{background_color}/{TEXT_COLOR}.png&text={text}"

    try:
        response = requests.get(url)
        response.raise_for_status()
        content = response.content
    except requests.RequestException as e:
        logger.error(f"Error fetching thumbnail: {e}")
        raise

    thumbnail_path = Path("src") / "posts" / "blog" / "images" / f"{file_name}.png"
    thumbnail_path.parent.mkdir(parents=True, exist_ok=True)

    try:
        thumbnail_path.write_bytes(content)
        logger.info(f"Thumbnail generated: {thumbnail_path}")
        return thumbnail_path
    except OSError as e:
        logger.error(f"Error writing thumbnail file: {e}")
        raise


def create_markdown_file(content: str, meta: Meta) -> None:
    current_date = datetime.now().strftime("%Y-%m-%d %H:%M")
    file_name = meta["meta"]["title"].replace(" ", "-").replace(":", "-").lower()

    try:
        thumbnail_path = generate_placeholder_thumbnail(file_name)
        relative_thumbnail_path = thumbnail_path.relative_to(
            Path("src") / "posts" / "blog"
        ).as_posix()

        file_path = Path("src") / "posts" / "blog" / f"{file_name}.md"

        markdown_content = f"""---
title: {meta["meta"]["title"].replace(":", "-")}
category: "constant"
date: {current_date} +00:00
desc: {meta["meta"]["description"].replace(":", "-")}
thumbnail: "./{relative_thumbnail_path}"
alt: "Thumbnail for {meta["meta"]["title"]}"
---

{content}"""

        file_path.parent.mkdir(parents=True, exist_ok=True)
        file_path.write_text(markdown_content, encoding="utf-8")

        logger.info(f"Markdown file created: {file_path}")
    except Exception as e:
        logger.error(f"Error creating markdown file: {e}")
        raise


def generate_single_content(
    high_temperature_model: BaseChatModel, low_temperature_model: BaseChatModel, topic: str
) -> None:
    try:
        outline = generate_outline(high_temperature_model, topic)
        content = generate_content(low_temperature_model, outline)
        meta = generate_meta(low_temperature_model, outline)
        create_markdown_file(content, meta)
    except Exception as e:
        logger.error(f"Error in content generation pipeline: {e}")


class Provider(StrEnum):
    openai = "openai"
    mistral = "mistral"


def get_chat_model(provider: Provider, temperature: float = 0.8) -> BaseChatModel:
    match provider:
        case Provider.openai:
            return ChatOpenAI(model="gpt-4o-mini", temperature=temperature)
        case Provider.mistral:
            return ChatMistralAI(model_name="open-mixtral-8x7b", temperature=temperature)


@app.command()
def main(
    n_content: int = typer.Option(
        1, "--n-content", "-n", help="Number of content pieces to generate"
    ),
    topic: str = typer.Option("free topic", "--topic", "-t", help="Topic for content generation"),
    provider: Annotated[Provider, typer.Option(case_sensitive=False)] = Provider.mistral,
) -> None:
    high_temperature_model = get_chat_model(provider, temperature=1.3)
    low_temperature_model = get_chat_model(provider, temperature=0.6)

    for i in range(n_content):
        logger.info(f"Generating content {i + 1} of {n_content}...")
        generate_single_content(high_temperature_model, low_temperature_model, topic)

    logger.info("Content generation complete.")


if __name__ == "__main__":
    app()
